### PR TITLE
RevWalk: push all parents and deduplicate later, return early if filled with unwanted

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have added myself to the contributors in `CHANGELOG.md` (optional)

### CODE REVIEW

**Review by martinvonz — COMMENTED** [review]
> Also, this appears to be faster than using boundary sets, based on the
bench extracted from test_index_commits_criss_cross().

Did you mean to include that in the patch?

**Comment by yuja** [comment]
> > Also, this appears to be faster than using boundary sets, based on the
> > bench extracted from test_index_commits_criss_cross().
> 
> Did you mean to include that in the patch?

No. It's an ugly copy-paste code I didn't intend to submit. I just wanted to make sure the new code is significantly worse than the original.
https://gist.github.com/yuja/392001099f1210a7acc58ad759a7be4f 

I think there should be a way to reuse the test code for benchmark, but I'm not familiar with the framework. And it seems the whole `cargo bench` stack isn't standardized yet?

File: lib/src/index.rs [hunk 1]
```diff
@@ -967,64 +967,77 @@ impl PartialOrd for IndexEntryByGeneration<'_> {
...
     type Item = IndexEntry<'a>;
 
     fn next(&mut self) -> Option<Self::Item> {
         while let Some(item) = self.items.pop() {
-            if item.wanted {
-                self.wanted_boundary_set.remove(&item.entry.0.pos);
-                if self.unwanted_boundary_set.contains(&item.entry.0.pos) {
-                    continue;
-                }
+            self.skip_while_eq(&item.entry.0);
```
Comment (martinvonz) [inline lib/src/index.rs:1046]: Are there cases where you would want to pop an item without also skipping equal elements? If not, it seems better to extract a helper for popping the next element and discarding duplicates.

Comment (yuja) [reply lib/src/index.rs:1046]: Ah, yes. I should've mentioned that in the commit message.
A RevWalk that supports generation range filtering will propagate every `Wanted(gen)` as `Wanted(gen + 1)`:
```rust
                while let Some(x) = self.queue.pop_eq(&item.entry.0) {
                    if let RevWalkWorkItemState::Wanted(gen) = x.state {
                        if known_gen != gen {
                            self.queue.push_wanted_parents(&item.entry.0, gen + 1);
                            known_gen = gen;
```

Comment (martinvonz) [reply lib/src/index.rs:1046]: Thanks for explaining. Sounds good to have it this way if it's planning for the future.

**Review by martinvonz — APPROVED** [review]

**Pull request merged**